evidence/store.go

```diff
@@ -0,0 +1,189 @@
+}
+
+// big endian padded hex
+func be(h int) string {
+	return fmt.Sprintf("%0.16X", h)
+}
+
+func keyLookupFromHeightAndHash(height int, hash []byte) []byte {
+	return _key("%s/%s/%X", baseKeyLookup, be(height), hash)
+}
+
+func keyOutqueue(evidence types.Evidence, priority int) []byte {
+	return _key("%s/%s/%s/%X", baseKeyOutqueue, be(priority), be(evidence.Height()), evidence.Hash())
+}
+
+func keyPending(evidence types.Evidence) []byte {
+	return _key("%s/%s/%X", baseKeyPending, be(evidence.Height()), evidence.Hash())
+}
+
+func _key(fmt_ string, o ...interface{}) []byte {
+	return []byte(fmt.Sprintf(fmt_, o...))
+}
+
+// EvidenceStore is a store of all the evidence we've seen, including
+// evidence that has been committed, evidence that has been verified but not broadcast,
+// and evidence that has been broadcast but not yet committed.
+type EvidenceStore struct {
+	db dbm.DB
+}
+
+func NewEvidenceStore(db dbm.DB) *EvidenceStore {
+	return &EvidenceStore{
+		db: db,
+	}
+}
+
+// PriorityEvidence returns the evidence from the outqueue, sorted by highest priority.
+func (store *EvidenceStore) PriorityEvidence() (evidence []types.Evidence) {
+	// reverse the order so highest priority is first
+	l := store.ListEvidence(baseKeyOutqueue)
+	l2 := make([]types.Evidence, len(l))
+	for i := range l {
+		l2[i] = l[len(l)-1-i]
+	}
+	return l2
+}
+
+// PendingEvidence returns all known uncommitted evidence.
+func (store *EvidenceStore) PendingEvidence() (evidence []types.Evidence) {
+	return store.ListEvidence(baseKeyPending)
+}
+
+// ListEvidence lists the evidence for the given prefix key.
+// It is wrapped by PriorityEvidence and PendingEvidence for convenience.
+func (store *EvidenceStore) ListEvidence(prefixKey string) (evidence []types.Evidence) {
+	iter := store.db.IteratorPrefix([]byte(prefixKey))
+	for iter.Next() {
+		val := iter.Value()
+
+		var ei EvidenceInfo
+		wire.ReadBinaryBytes(val, &ei)
+		evidence = append(evidence, ei.Evidence)
+	}
+	return evidence
+}
+
+// GetEvidence fetches the evidence with the given height and hash.
+func (store *EvidenceStore) GetEvidence(height int, hash []byte) *EvidenceInfo {
+	key := keyLookupFromHeightAndHash(height, hash)
+	val := store.db.Get(key)
+
+	if len(val) == 0 {
+		return nil
+	}
+	var ei EvidenceInfo
+	wire.ReadBinaryBytes(val, &ei)
+	return &ei
+}
+
+// AddNewEvidence adds the given evidence to the database.
+// It returns false if the evidence is already stored.
+func (store *EvidenceStore) AddNewEvidence(evidence types.Evidence, priority int) bool {
+	// check if we already have seen it
+	ei_ := store.GetEvidence(evidence.Height(), evidence.Hash())
+	if ei_ != nil && ei_.Evidence != nil {
+		return false
+	}
+
+	ei := EvidenceInfo{
+		Committed: false,
+		Priority:  priority,
+		Evidence:  evidence,
+	}
+	eiBytes := wire.BinaryBytes(ei)
+
+	// add it to the store
+	key := keyOutqueue(evidence, priority)
+	store.db.Set(key, eiBytes)
```

Comment:
I wonder if the same problem applies here `If we crash after committed but before removing/updating` (i.e. if one of the set op crashes)

Reply:
These operations seem idempotent... see my other comment on idempotency